posts/aws-appsync-is-coming-for-your-api.md

```diff
@@ -0,0 +1,125 @@
+
+![AWS AppSync Overview](https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2018/06/appSync_overview.png)
+
+At a high level, AppSync exposes a [GraphQL-based](https://graphql.org/) API, which then allows you to resolve your  GraphQL fields to various backend services, which currently includes DynamoDB, Lambda and ElasticSearch. The other magic sauce is the real-time websockets integration (presumably using [Amazon IoT](https://github.com/aws/aws-iot-device-sdk-js/blob/master/README.md#browser) under the hood), which pushes changes to fields back to your clients (if they've subscribed to changes). 
+
+For any non-trivial implementation, you'll probably want to use [Amazon Cognito](https://aws.amazon.com/cognito/) to handle the authentication aspects of your application. Authorisation is enforced by AppSync, and can use either API key, AWS IAM, or a Congito User pool for authentication.
+
+# Hacking Around
+![AWS AppSync Console Screenshot](https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2018/06/appSyncConsole.png)
+
+The GraphQL schema is the beating heart of AppSync. GraphQL is very powerful (particularly for developing complex cross-queries), but it takes a while to get your head around it. I've put together a super-simple example based on a rudimentary vote client to demonstrate the primary concepts.
+
+```graph
+type Candidate {
+	candidateId: ID!
+	votes: Int!
+}
+
+type Mutation {
+	vote(candidateId: ID!): Candidate
+}
+
+type Query {
+	getCandidates: [Candidate]
+}
+
+type Subscription {
+	onUpdateCandidate(candidateId: ID, votes: Int): Candidate
+		@aws_subscribe(mutations: ["vote"])
+}
+
+schema {
+	query: Query
+	mutation: Mutation
+	subscription: Subscription
+}
+```
+
+The most confusing part of the schema definition is the `schema` special type, which must have a `query` type attribute, and may have a `mutation` type. In additional, AppSync uses a `subscription` type to indicate which data will need to be pushed back to subscribed clients. Once the schema has been successfully loaded (it has inbuild validation), the next step is to "Create Resources". This allows you to automagically create DynamoDB tables based on the schema. It will also extend your GraphQL schema to add common mutations and queries to your schema.
+
+Once a Data Source has been added, you can then attaching custom Resolvers through the schema page. This entire setup reminds me of the old Lambda mapping behaviour, but I'm sure it'll improve over time. The resolver is a [Velocity Template](http://velocity.apache.org/engine/1.7/user-guide.html), which you use to generate a JSON request to the backend service, in this instance, DynamoDB. I can see people just ignoring them and going straight to Lambda as a proxy. To demonstrate the complexity within the template, following request-mapper is used to increment the vote counter for the `vote` mutation.
```

Comment:
"I can see people just ignoring them and going straight to Lambda as a proxy." Why? How does that link to the rest of the paragraph?

Reply:
I've moved this down so it flows better.